From 72071f282c28f4b3d5bb187d3d85fee1ccc6e20a Mon Sep 17 00:00:00 2001
From: Danny Lin <danny@kdrag0n.dev>
Date: Tue, 5 Oct 2021 20:25:19 -0700
Subject: [PATCH 002/103] pm: Add seamless migration between test and release
 keys

Official release builds should be signed with release keys for security,
but switching from test keys forces users to wipe data because the
system will no longer boot due to signature mismatches.

To improve the security of release builds while avoiding unnecessary
data loss, seamlessly migrate packages between different groups of
platform keys (test -> release for users, and release -> test for
development) at boot time.

Key migration needs to be configured with the following system
properties, which should be set to the signing certificates (*.x509.pem)
encoded as hex strings (not base64 or base64 encoded as hex):

ro.build.certs.release.default
ro.build.certs.release.media
ro.build.certs.release.networkstack
ro.build.certs.release.platform
ro.build.certs.release.shared
ro.build.certs.test.default
ro.build.certs.test.media
ro.build.certs.test.networkstack
ro.build.certs.test.platform
ro.build.certs.test.shared

[ YumeMichi: Add sdk_sandbox for A15. ]

Change-Id: Ic50bef066ca49d0a1efb42a6bbb99c142865bb1d
---
 .../server/pm/KeySetManagerService.java       |   3 +-
 .../android/server/pm/PackageSignatures.java  |   3 +-
 .../server/pm/PlatformKeyMigration.java       | 112 ++++++++++++++++++
 3 files changed, 116 insertions(+), 2 deletions(-)
 create mode 100644 services/core/java/com/android/server/pm/PlatformKeyMigration.java

diff --git a/services/core/java/com/android/server/pm/KeySetManagerService.java b/services/core/java/com/android/server/pm/KeySetManagerService.java
index f1453c874faf..79f8226a4105 100644
--- a/services/core/java/com/android/server/pm/KeySetManagerService.java
+++ b/services/core/java/com/android/server/pm/KeySetManagerService.java
@@ -814,7 +814,8 @@ public class KeySetManagerService {
             throws XmlPullParserException {
         long identifier = parser.getAttributeLong(null, "identifier");
         int refCount = 0;
-        byte[] publicKey = parser.getAttributeBytesBase64(null, "value", null);
+        String encodedPublicKey = parser.getAttributeValue(null, "value");
+        byte[] publicKey = PlatformKeyMigration.mapKey(encodedPublicKey);
         PublicKey pub = FrameworkParsingPackageUtils.parsePublicKey(publicKey);
         if (pub != null) {
             PublicKeyHandle pkh = new PublicKeyHandle(identifier, refCount, pub);
diff --git a/services/core/java/com/android/server/pm/PackageSignatures.java b/services/core/java/com/android/server/pm/PackageSignatures.java
index 90f57a7b74ba..1d65143471af 100644
--- a/services/core/java/com/android/server/pm/PackageSignatures.java
+++ b/services/core/java/com/android/server/pm/PackageSignatures.java
@@ -169,7 +169,8 @@ class PackageSignatures {
                     if (index != -1) {
                         boolean signatureParsed = false;
                         try {
-                            final byte[] key = parser.getAttributeBytesHex(null, "key", null);
+                            final String rawKey = parser.getAttributeValue(null, "key");
+                            final byte[] key = PlatformKeyMigration.mapCert(rawKey);
                             if (key == null) {
                                 if (index >= 0 && index < readSignatures.size()) {
                                     Signature sig = readSignatures.get(index);
diff --git a/services/core/java/com/android/server/pm/PlatformKeyMigration.java b/services/core/java/com/android/server/pm/PlatformKeyMigration.java
new file mode 100644
index 000000000000..658f1eff193e
--- /dev/null
+++ b/services/core/java/com/android/server/pm/PlatformKeyMigration.java
@@ -0,0 +1,112 @@
+/*
+ * Copyright (C) 2021 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.server.pm;
+
+import android.content.pm.Signature;
+import android.os.Build;
+import android.os.SystemProperties;
+import android.util.Base64;
+
+import com.android.internal.util.HexDump;
+
+import java.security.cert.CertificateException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+final class PlatformKeyMigration {
+    // We spoof tags for SafetyNet, but not the display build ID
+    private static final boolean IS_RELEASE = !Build.DISPLAY.contains("test-keys");
+
+    private static final List<String> KEY_NAMES = Arrays.asList(
+            "media",
+            "networkstack",
+            "platform",
+            "default", // releasekey/testkey
+            "shared",
+            "sdk_sandbox"
+            /* AVB key doesn't matter here */
+    );
+
+    private static final Map<String, String> replacementMap = createReplacementMap();
+
+    private static Map<String, Signature> loadCertsFromProps(String group) {
+        Map<String, Signature> certs = new HashMap<>();
+        for (String name : KEY_NAMES) {
+            String encodedCert = SystemProperties.get("ro.build.certs." + group + "." + name);
+            if (encodedCert == null || encodedCert.equals("")) {
+                continue;
+            }
+
+            Signature cert = new Signature(encodedCert);
+            certs.put(name, cert);
+        }
+
+        return certs;
+    }
+
+    private static String certToPublicKey(Signature cert) throws CertificateException {
+        return Base64.encodeToString(cert.getPublicKey().getEncoded(), Base64.NO_WRAP);
+    }
+
+    private static Map<String, String> buildMappings(Map<String, Signature> from, Map<String, Signature> to)
+            throws CertificateException {
+        Map<String, String> replacementMap = new HashMap<>();
+        for (Map.Entry<String, Signature> fromEntry : from.entrySet()) {
+            String name = fromEntry.getKey();
+            Signature fromCert = fromEntry.getValue();
+            Signature toCert = to.get(name);
+
+            // Forward mapping only, since this is directional
+            replacementMap.put(fromCert.toCharsString(), toCert.toCharsString());
+            replacementMap.put(certToPublicKey(fromCert), certToPublicKey(toCert));
+        }
+
+        return replacementMap;
+    }
+
+    private static Map<String, String> createReplacementMap() {
+        Map<String, Signature> releaseCerts = loadCertsFromProps("release");
+        Map<String, Signature> testCerts = loadCertsFromProps("test");
+
+        try {
+            if (IS_RELEASE) {
+                // User test -> release migration
+                return buildMappings(testCerts, releaseCerts);
+            } else {
+                // Dev release -> test migration
+                return buildMappings(releaseCerts, testCerts);
+            }
+        } catch (CertificateException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static String mapValue(String value) {
+        String replacement = replacementMap.get(value);
+        return replacement != null ? replacement : value;
+    }
+
+    public static byte[] mapCert(String cert) {
+        return cert != null ? HexDump.hexStringToByteArray(mapValue(cert)) : null;
+    }
+
+    public static byte[] mapKey(String key) {
+        return key != null ? Base64.decode(mapValue(key), 0) : null;
+    }
+}
-- 
2.47.1

